File: back-end/src/main/java/es/ull/project/application/usecase/vehicle/UpdateVehicleUseCase.java
package es.ull.project.application.usecase.vehicle;

import java.util.UUID;

import es.ull.project.domain.entity.Vehicle;
import es.ull.project.domain.enumerate.VehicleType;
import es.ull.project.domain.valueobject.capacity.VehicleCapacityKilograms;
import es.ull.project.domain.valueobject.capacity.VehicleCapacityLiters;
import es.ull.project.domain.valueobject.cost.TransportationVariableCost;

/**
 * Use case for updating a vehicle.
 */
public interface UpdateVehicleUseCase {
    /**
     * Updates a vehicle with the specified attributes.
     *
     * @param id                     the unique identifier of the vehicle to update
     * @param newVehicleType         the new vehicle type
     * @param newCapacityKilograms  the new capacity in kilograms
     * @param newCapacityLiters      the new capacity in liters
     * @param newCostPerKilometer    the new cost per kilometer
     * @return the updated vehicle
     */
    Vehicle update(UUID id, VehicleType newVehicleType, VehicleCapacityKilograms newCapacityKilograms, VehicleCapacityLiters newCapacityLiters, TransportationVariableCost newCostPerKilometer);
}

File: front-end/src/application/usecase/VehicleManagement/UpdateVehicleUseCase.ts
// UpdateVehicleUseCase.ts
// Use case contract for updating a vehicle

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { Vehicle } from '../../../domain/entity/Vehicle';
import { VehicleType } from '../../../domain/enumerate/VehicleType';
import { Capacity } from '../../../domain/valueobject/demand/Capacity';
import { TransportationVariableCost } from '../../../domain/valueobject/cost/TransportationVariableCost';

/**
 * Use case for updating an existing vehicle in the system.
 * Input: vehicleId and fields to update
 * Output: Vehicle (entity)
 */

// Command object for input data
export interface UpdateVehicleCommand {
    vehicleId: string;
    updatedFields: Partial<{
        vehicleType: VehicleType;
        transportCapacity: Capacity;
        costPerKilometer: TransportationVariableCost;
    }>;
}

// Result type for the use case
export type UpdateVehicleResult = Vehicle;

// Use case contract
export interface UpdateVehicleUseCase {
    /**
     * Handles the update of an existing vehicle
     *
     * @param command Data needed to update the vehicle
     * @returns Either a DataError or the updated Vehicle
     */
    execute(command: UpdateVehicleCommand): Promise<Either<DataError, UpdateVehicleResult>>;
}
